sklearn/tree/_export.py

```diff
@@ -714,6 +720,21 @@ def recurse(self, node, tree, ax, max_x, max_y, depth=0):
                     (max_y - node.parent.y - 0.5) / max_y,
                 )
                 ax.annotate(node.tree.label, xy_parent, xy, **kwargs)
+
+                # Draw True/False labels if parent is root node
+                if node.parent.parent is None:
+                    # Adjust the position for the text to be slightly above the arrow
+                    text_pos = (
+                        (xy_parent[0] + xy[0]) / 2 + 0.05,
+                        (xy_parent[1] + xy[1]) / 2,
+                    )
```

Comment:
I'm thinking about using `ha="right"` for the left branch and `ha="left"` for the left branch instead of hardcoding an offset (which can cause the label to be very far away from the arrow for certain figure sizes). You may then need to refactor the `non_box_kwargs` and `kwargs` part a bit.

Reply:
Maybe like this:

```python
text_pos = (
    (xy_parent[0] + xy[0]) / 2,
    (xy_parent[1] + xy[1]) / 2,
)
# Annotate the arrow with the edge label to indicate the child
# where the sample-split condition is satisfied
if node.parent.left() == node:
    label_text, label_ha = ("True ", "right")
else:
    label_text, label_ha = (" False", "left")
ax.annotate(label_text, text_pos, ha=label_ha, **non_box_kwargs)
```

When `node.parent.left() == node` we set `ha="right"` so that the rightmost of the label text will be at the center of the arrow, i.e., the text will be to the left of the arrow. The space in `"True "` is to create a small offset from the arrow to avoid overlapping. Not sure what other maintainers think about this approach. Similar case for the other branch.

Note that `ha="center"` will then need to be removed from `non_box_kwargs` and added only for `kwargs`.

Reply:
This does not work unfortunatlely. 

![Figure_1](https://github.com/scikit-learn/scikit-learn/assets/3460267/a18179ac-d5d4-41a6-a67c-7e19b8be0c17)

The `ha='center'` actually always gets the label very close to the midpoint of the arrow. If you try the code I pasted in https://github.com/scikit-learn/scikit-learn/pull/28552#issuecomment-1969989803, it seems robust.

Reply:
Maybe there are some misunderstandings. This is what I get from running the script you provided:

![aaa](https://github.com/scikit-learn/scikit-learn/assets/108576690/1812216a-846e-4c18-9460-4ec6668c49b5)

The modifications I make in the kwargs setting part:

```python
        # kwargs for annotations without a bounding box
        non_box_kwargs = dict(
            va="center",
            zorder=100 - 10 * depth,
            xycoords="axes fraction",
        )
        if self.fontsize is not None:
            non_box_kwargs["fontsize"] = self.fontsize

        # kwargs for annotations with a bounding box
        kwargs = dict(
            ha="center",
            arrowprops=self.arrow_args.copy(),
            bbox=self.bbox_args.copy(),
            **non_box_kwargs,
        )
        kwargs["arrowprops"]["edgecolor"] = plt.rcParams["text.color"]
```

The modifications I make in the label drawing part:

```python
                # Draw True/False labels if parent is root node
                if node.parent.parent is None:
                    # Adjust the position for the text to be slightly above the arrow
                    text_pos = (
                        (xy_parent[0] + xy[0]) / 2,
                        (xy_parent[1] + xy[1]) / 2,
                    )
                    # Annotate the arrow with the edge label to indicate the child
                    # where the sample-split condition is satisfied
                    if node.parent.left() == node:
                        label_text, label_ha = ("True  ", "right")
                    else:
                        label_text, label_ha = ("  False", "left")
                    ax.annotate(label_text, text_pos, ha=label_ha, **non_box_kwargs)
```

Reply:
Do you need me to make explicit suggestions based on the current diff or the code snippet above is enough?

Reply:
Sorry I meant that when changing the code to something like https://github.com/scikit-learn/scikit-learn/pull/28552#discussion_r1510258103, I get the result linked in https://github.com/scikit-learn/scikit-learn/pull/28552#discussion_r1510289747. 

The original code with `ha='center'` is robust and works well when changing figure sizes versus the suggested code seems to make the True/False appear weirdly. 

WDYT?

Reply:
The current version may not look that well when the figure is wide.

<details>
<summary>A code snippet I modified based on yours</summary>
<p></p>

```python
import numpy as np
from matplotlib import pyplot as plt

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree

iris = load_iris()
X = iris.data
y = iris.target
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

clf = DecisionTreeClassifier(max_leaf_nodes=6, random_state=0)
clf.fit(X_train, y_train)

plt.figure(figsize=(20, 8))
tree.plot_tree(
    clf,
    feature_names=iris.feature_names,
    class_names=iris.target_names,
    fontsize=8,
    filled=True,
)
plt.savefig("aaa.png")
```

</details>

The current version produces:

![aaa](https://github.com/scikit-learn/scikit-learn/assets/108576690/c0c8501a-4b2a-4ab6-a3ec-f41e6c63f892)

My suggested version produces:

![aaa](https://github.com/scikit-learn/scikit-learn/assets/108576690/aaa58043-f730-49ae-b0e8-6b3b0d393385)

Reply:
I confirm I get the good looking graph as you suggested!